# Metody Inteligencji Obliczeniowej w Analizie Danych

[Tasks](https://pages.mini.pw.edu.pl/~karwowskij/mioad/lab-sieci.html)

### MLP Lab 1

Na tych zajęciach należy zaimplementować sieć neuronową typu MLP, w której można ustawić: liczbę warstw, liczbę neuronów w każdej z warstw i wagi poszczególnych połączeń, w tym biasów. Sieć ma **używać sigmoidalnej funkcji** aktywacji. **Na wyjściu** dopuszczana jest funkcja **liniowa**.

Implementacja sieci musi być przygotowana w taki sposób, żeby łatwo zmieniać:

- Architekturę, to znaczy liczbę wejść, wyjść, neuronów w warstwach ukrytych. 
- Funkcję aktywacji.  

Tak przygotowaną implementację należy następnie wykorzystać do rozwiązania zadania regresji na dostarczonych danych. Parametry sieci należy dobrać ręcznie, tak aby uzyskać możliwie dobrze wyniki na zbiorach danych (zbudować po jednej sieci dla każdego zbioru):

square-simple
steps-large
Rozważyć architektury sieci:

- jedna warstwa ukryta, 5 neuronów,  
- jedna warstwa ukryta, 10 neuronów,
- dwie warstwy ukryte, po 5 neuronów każda.

Aby otrzymać 2 punkty MSE na [nieznormalizowanym] zbiorze testowym nie może przekraczać wartości 9.

#### TODO:

- add asserts
- create more tests

In [1]:
from MultiLayerPerceptron import NeuralNetwork, Layer
from itertools import chain
import pandas as pd
import numpy as np

def mse(real, pred):
    return np.square(np.subtract(real,pred)).mean() 

## Dataset 2: Square Simple

In [2]:
df_train = pd.read_csv('data/regression/square-simple-training.csv').set_index("Unnamed: 0")
df_test = pd.read_csv('data/regression/square-simple-test.csv').set_index("Unnamed: 0")
print(df_train.head())

x_train = [[x] for x in df_train.loc[:,"x"]]
y_train = [y for y in df_train.loc[:,"y"]]
x_test = [[x] for x in df_test.loc[:,"x"]]
y_test = [y for y in df_test.loc[:,"y"]]

                   x           y
Unnamed: 0                      
1          -0.171543 -127.351580
2           0.025201 -129.942844
3          -1.368991   38.672367
4           1.907390  197.432191
5           0.011129 -129.988852


#### Model 1: 1 hidden layer with 5 neurons

In [3]:
net1_1_5_1 = NeuralNetwork(weights_random=False)
net1_1_5_1.add(Layer(neurons_count=1, add_bias=True))
net1_1_5_1.add(Layer(neurons_count=5, activation_fun="sigmoid", add_bias=True))
net1_1_5_1.add(Layer(neurons_count=1, activation_fun="linear", add_bias=False))
net1_1_5_1.set_weights(weights=[\
        np.array([[-2.9,-2.8897,-2.8897,-2.8897,-2.8897],\
                [-0.9733,-0.9733,-0.9733,-0.9733,0.9733]]),\
        np.array([[-339.16],\
                [502],\
                [502],\
                [502],\
                [502],\
                [1993]])])
print(net1_1_5_1)

Neural network layers:
	Layer 1: Layer has 2 neurons (including 1 bias neuron) and activation function is 'linear function'
	Layer 2: Layer has 6 neurons (including 1 bias neuron) and activation function is 'sigmoid function'
	Layer 3: Layer has 1 neurons (with no bias neuron) and activation function is 'linear function'
Neural network weights:
	Weights 1: (2, 5) (input, output)
[[-2.9    -2.8897 -2.8897 -2.8897 -2.8897]
 [-0.9733 -0.9733 -0.9733 -0.9733  0.9733]]
	Weights 2: (6, 1) (input, output)
[[-339.16]
 [ 502.  ]
 [ 502.  ]
 [ 502.  ]
 [ 502.  ]
 [1993.  ]]



In [4]:
y_train_pred_1 = net1_1_5_1.predict(x_train)
y_train_pred_1 = list(chain.from_iterable(y_train_pred_1))
mse_1_1 = mse(y_train_pred_1, y_train)
print(f"MSE train is {round(mse_1_1,2)}")

MSE train is 2.46


In [5]:
y_test_pred_1 = net1_1_5_1.predict(x_test)
y_test_pred_1 = list(chain.from_iterable(y_test_pred_1))
mse_1_2 = mse(y_test_pred_1, y_test)
print(f"MSE test is {round(mse_1_2,2)}")

MSE test is 2.26


#### Model 2: 1 hidden layer with 10 neurons

In [6]:
net2_1_10_1 = NeuralNetwork(weights_random=False)
net2_1_10_1.add(Layer(neurons_count=1, add_bias=True))
net2_1_10_1.add(Layer(neurons_count=10, activation_fun="sigmoid", add_bias=True))
net2_1_10_1.add(Layer(neurons_count=1, activation_fun="linear", add_bias=False))
net2_1_10_1.set_weights(weights=[\
        np.array([[-2.8897,-2.8897,-2.8897,-2.8897,-2.8897,-2.8897,-2.8897,-2.8897,-2.8897,-2.8897],\
                [-0.9733,-0.9733,-0.9733,-0.9733,-0.9733,0.9733,0.9733,0.9733,0.9733,0.9733]]),\
        np.array([[-339.3],\
                [400],\
                [400],\
                [400],\
                [400],\
                [400],\
                [400],\
                [400],\
                [400],\
                [400],\
                [400]])])
print(net2_1_10_1)

Neural network layers:
	Layer 1: Layer has 2 neurons (including 1 bias neuron) and activation function is 'linear function'
	Layer 2: Layer has 11 neurons (including 1 bias neuron) and activation function is 'sigmoid function'
	Layer 3: Layer has 1 neurons (with no bias neuron) and activation function is 'linear function'
Neural network weights:
	Weights 1: (2, 10) (input, output)
[[-2.8897 -2.8897 -2.8897 -2.8897 -2.8897 -2.8897 -2.8897 -2.8897 -2.8897
  -2.8897]
 [-0.9733 -0.9733 -0.9733 -0.9733 -0.9733  0.9733  0.9733  0.9733  0.9733
   0.9733]]
	Weights 2: (11, 1) (input, output)
[[-339.3]
 [ 400. ]
 [ 400. ]
 [ 400. ]
 [ 400. ]
 [ 400. ]
 [ 400. ]
 [ 400. ]
 [ 400. ]
 [ 400. ]
 [ 400. ]]



In [7]:
y_train_pred_2 = net2_1_10_1.predict(x_train)
y_train_pred_2 = list(chain.from_iterable(y_train_pred_2))
mse_1_2 = mse(y_train_pred_2, y_train)
print(f"MSE train is {round(mse_1_2,2)}")

MSE train is 3.11


In [8]:
y_test_pred_2 = net2_1_10_1.predict(x_test)
y_test_pred_2 = list(chain.from_iterable(y_test_pred_2))
mse_1_2 = mse(y_test_pred_2, y_test)
print(f"MSE test is {round(mse_1_2,2)}")

MSE test is 2.63


## Dataset 2: Steps Large

In [9]:
df_train = pd.read_csv('data/regression/steps-large-training.csv').set_index("Unnamed: 0")
df_test = pd.read_csv('data/regression/steps-large-test.csv').set_index("Unnamed: 0")
print(df_train.head())

x_train = [[x] for x in df_train.loc[:,"x"]]
y_train = [y for y in df_train.loc[:,"y"]]
x_test = [[x] for x in df_test.loc[:,"x"]]
y_test = [y for y in df_test.loc[:,"y"]]

                   x   y
Unnamed: 0              
1          -1.481354 -80
2           1.033264  80
3          -0.076403   0
4          -1.419785 -80
5          -0.108398   0


#### Model 1: 2 hidden layers with 5 neurons each

In [10]:
net3_1_5_1 = NeuralNetwork()
net3_1_5_1.add(Layer(neurons_count=1, add_bias=True))
net3_1_5_1.add(Layer(neurons_count=5, activation_fun='sigmoid', add_bias=True))
net3_1_5_1.add(Layer(neurons_count=1, activation_fun='linear', add_bias=False))

b = -250
a = 500
c = -80
d = 80
e = 80/2

net3_1_5_1.set_weights([\
        np.array([[-250,-750,-750,250,250],\
                [501,499.95,499.95,500.2,500.2]]),\
        np.array([[-80],\
                [80],\
                [40],\
                [40],\
                [40],\
                [40]])])
print(net3_1_5_1)

Neural network layers:
	Layer 1: Layer has 2 neurons (including 1 bias neuron) and activation function is 'linear function'
	Layer 2: Layer has 6 neurons (including 1 bias neuron) and activation function is 'sigmoid function'
	Layer 3: Layer has 1 neurons (with no bias neuron) and activation function is 'linear function'
Neural network weights:
	Weights 1: (2, 5) (input, output)
[[-250.   -750.   -750.    250.    250.  ]
 [ 501.    499.95  499.95  500.2   500.2 ]]
	Weights 2: (6, 1) (input, output)
[[-80]
 [ 80]
 [ 40]
 [ 40]
 [ 40]
 [ 40]]



In [11]:
y_train_pred_2 = net3_1_5_1.predict(x_train)
y_train_pred_2 = list(chain.from_iterable(y_train_pred_2))
mse_1_2 = mse(y_train_pred_2, y_train)
print(f"MSE train is {round(mse_1_2,2)}")

MSE train is 4.14


In [12]:
y_test_pred_2 = net3_1_5_1.predict(x_test)
y_test_pred_2 = list(chain.from_iterable(y_test_pred_2))
mse_1_2 = mse(y_test_pred_2, y_test)
print(f"MSE test is {round(mse_1_2,2)}")

MSE test is 3.58


#### Results

MSE for the `square simple` dataset is:  
- train: 2.46
- test: 2.26

MSE for the `steps large` dataset is:  
- train: 4.14
- test: 3.58
